include/oneapi/dpl/pstl/hetero/dpcpp/utils_ranges_sycl.h

```diff
@@ -45,6 +45,36 @@ __create_accessor(_BufferType& __buf, _DiffType __offset, _DiffType __n)
 
     return {__buf, sycl::range<1>(__n_acc), __offset};
 }
+
+// Evaluates to true if the provided type is an iterator with a value_type and if the implementation of a
+// std::vector<value_type, Alloc>::iterator can be distinguished between three different allocators, the
+// default, usm_shared, and usm_host. If all are distinct, it is very unlikely any non-usm based allocator
+// could be confused with a usm allocator.
+template <typename Iter, typename Void = void>
+struct __vector_iter_distinguishes_by_allocator : ::std::false_type
+{
+};
+
+template <typename Iter, typename ValueType = typename ::std::iterator_traits<Iter>::value_type>
+using __default_alloc_vec_iter = typename ::std::vector<ValueType>::iterator;
+
+template <typename Iter, typename ValueType = typename ::std::iterator_traits<Iter>::value_type>
+using __usm_shared_alloc_vec_iter =
+    typename ::std::vector<ValueType, typename sycl::usm_allocator<ValueType, sycl::usm::alloc::shared>>::iterator;
+
+template <typename Iter, typename ValueType = typename ::std::iterator_traits<Iter>::value_type>
+using __usm_host_alloc_vec_iter =
+    typename ::std::vector<ValueType, typename sycl::usm_allocator<ValueType, sycl::usm::alloc::host>>::iterator;
+
+template <typename Iter>
+struct __vector_iter_distinguishes_by_allocator<
```

Comment:
I guess we played it extra safe here, but let be so.

Reply:
Yes, but this provides more and better evidence that unique allocators (not defined by the standard) result in iterator types which are unique from each other.  All of it is done at compile time, so I don't see a downside.